### PR TITLE
fix(sqlparser): array of struct of struct

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

When parsing `>>[]`, whose token stream contains `[ShiftRight, LBracket, RBracket]`, the parser first consumes `ShiftRight` and sets internal state `remaining_close = true`. At this stage, it shall consider the next token as effectively `Gt` rather than peeking at `LBracket` from the stream. In the latter case it would be wrongly parsed as `>[]>`

[dot](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20%20%20struct_data_type%0A%20%20%20%20data_type%0A%20%20%20%20data_type_stateful%0A%20%20%20%20data_type_stateful_inner%0A%20%20%20%20keyword_datatype%0A%20%20%20%20non_keyword_datatype%0A%20%20%20%20map_type_arguments%0A%20%20%20%20%0A%20%20%20%20data_type%20-%3E%20data_type_stateful%0A%20%20%20%20data_type_stateful%20-%3E%20data_type_stateful_inner%0A%20%20%20%20data_type_stateful_inner%20-%3E%20keyword_datatype%0A%20%20%20%20data_type_stateful_inner%20-%3E%20non_keyword_datatype%0A%20%20%20%20keyword_datatype%20-%3E%20struct_data_type%0A%20%20%20%20non_keyword_datatype%20-%3E%20map_type_arguments%0A%20%20%20%20map_type_arguments%20-%3E%20keyword_datatype%0A%20%20%20%20map_type_arguments%20-%3E%20data_type_stateful%0A%20%20%20%20struct_data_type%20-%3E%20data_type_stateful%0A%7D)

Limitations:
* Ideally all places peeking from stream need to check `remaining_close` first as if it was the next token. But I cannot find a good way to ensure such a *virtual stream* is followed (mutating input stream is not an option as it breaks backtrack and location reporting). The current fix only applies to array syntax `[]` as suffix. Hopefully no future syntax extensions hit the same problem.

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [x] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/sqlparser/tests/testdata/array.yaml [hunk 1]
```diff
@@ -74,3 +74,6 @@
     sql parser error: expected an expression, found: [
     LINE 1: SELECT [1,2]
                    ^
+- input: CREATE TABLE t (params STRUCT<a STRUCT<b int>>[])
```
Comment (xxchan) [inline src/sqlparser/tests/testdata/array.yaml:77]: This test case seems to work in current version. Did I miss something?
```
dev=> CREATE TABLE t (params STRUCT<a STRUCT<b int>>[])
dev-> ;
CREATE_TABLE
dev=> describe t;
┌───────────────────┬─────────────────────┬───────────┬─────────────┐
│       Name        │        Type         │ Is Hidden │ Description │
├───────────────────┼─────────────────────┼───────────┼─────────────┤
│ params            │                     │ false     │ ∅           │
│ params.a          │ struct<b integer>[] │ false     │ ∅           │
│ _row_id           │ serial              │ true      │ ∅           │
│ primary key       │ _row_id             │ ∅         │ ∅           │
│ distribution key  │ _row_id             │ ∅         │ ∅           │
│ table description │ t                   │ ∅         │ ∅           │
└───────────────────┴─────────────────────┴───────────┴─────────────┘
(6 rows)
```

Comment (xxchan) [reply src/sqlparser/tests/testdata/array.yaml:77]: Oh, it's wrong

Comment (xxchan) [reply src/sqlparser/tests/testdata/array.yaml:77]: How did you find the bug? Is it related with other bugs?

Comment (xiangjinwu) [reply src/sqlparser/tests/testdata/array.yaml:77]: See the linked issue. It was causing a schema mismatch panic when `sink into table`.

Comment (xxchan) [reply src/sqlparser/tests/testdata/array.yaml:77]: Oh, I didn't notice there's linked issue since there's no `fix #xx` in body 😄

**Review by TennyZhuang — APPROVED** [review]
LGTM

**Review by xxchan — APPROVED** [review]
LGTM for the fix. But agree that not sure whether it's universal.

**Comment by xxchan** [comment]
> dot

Curious how did you generated this graph and what did you use it for? :eyes:

**Comment by xiangjinwu** [comment]
> > dot
> 
> Curious how did you generated this graph and what did you use it for? 👀

* Generated manually ...
* Used it to understand the indirect recursion, specially how parsing of `>>[]` became `>[]>`.

**Review by BugenZhao — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**